### PR TITLE
feat: [gd] persist ppw extra state

### PR DESCRIPTION
**Pull request by lalo**
This is a proper fix for the problem patched in https://github.com/VowpalWabbit/vowpal_wabbit/pull/3998 but for gd only.
Also related: https://github.com/VowpalWabbit/vowpal_wabbit/pull/4020

- adds an std::vector holding instances of state that is per model, i.e. set by ppw when building the reduction stack
- adds logic to swap state based on the incoming ft_offset
- learner.h allows to register a function used to resize the std::vector, since the base reduction has already instantiated by the time another reduction is registering their own ppw. The resizing calls are recursive to make sure all the stack is aware of the change.
- todo: reset particular offset

### CODE REVIEW

File: test/pred-sets/ref/ccb_quad.inv [hunk 1]
```diff
@@ -30,6 +30,9 @@ sd::oec.weighted_labeled_examples 8
 current_pass 1
 l1_state 0
 l2_state 1
+gd_ppw_state.size() = 1
+gd_ppw_state[0]normalized_sum_norm_x = 305.4242420196533
```
Comment (lalo) [inline test/pred-sets/ref/ccb_quad.inv:34]: right now this info is duplicated to maintain compat with previous versions see:

norm normalizer 305.424
total_weight 15.2712

We could skip duplicating this info when ppw_state.size() is equal to 1

File: vowpalwabbit/core/include/vw/core/learner.h [hunk 2]
```diff
@@ -710,7 +727,11 @@ struct reduction_learner_builder
   {
     this->_learner->weights = params_per_weight;
     this->_learner->increment = this->_learner->learn_fd.base->increment * this->_learner->weights;
-    if (params_per_weight > 1) { this->_learner->learn_fd.base->resize_ppw_state(params_per_weight); }
+    if (params_per_weight > 1)
+    {
+      this->_learner->learn_fd.base->resize_ppw_state(params_per_weight, this->_learner->increment);
+      this->_learner->MAX_FT_OFFSET = this->_learner->increment;
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/learner.h:733]: Why all caps? It isnt a constant

File: vowpalwabbit/core/include/vw/core/learner.h [hunk 3]
```diff
@@ -159,6 +159,14 @@ inline void debug_decrement_depth(multi_ex& ec_seq)
   }
 }
 
+inline size_t get_offset(example& ex) { return ex.ft_offset; }
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/learner.h:162]: const

File: version.txt [hunk 4]
```diff
@@ -1 +1 @@
-9.1.0
\ No newline at end of file
+99.1.0
```
Comment (bassmang) [inline version.txt:1]: typo?

File: vowpalwabbit/core/include/vw/core/vw_versions.h [hunk 5]
```diff
@@ -63,5 +63,8 @@ constexpr VW::version_struct VERSION_FILE_WITH_ACTIVE_SEEN_LABELS{9, 0, 0};
 /// Moved option values from command line to model data
 constexpr VW::version_struct VERSION_FILE_WITH_L1_AND_L2_STATE_IN_MODEL_DATA{9, 0, 0};
 
+/// GD, freegrad, ftrl are able to persist the correct state if ppw is higher than 1
+constexpr VW::version_struct VERSION_FILE_WITH_GD_PPW_STATE{99, 0, 0};
```
Comment (bassmang) [inline vowpalwabbit/core/include/vw/core/vw_versions.h:67]: 9.1.0?

File: version.txt (hunk 4 above)
Comment (lalo) [reply version.txt:1]: nope, its just a dummy future version since this model change needs a special version from the get-go

File: vowpalwabbit/core/include/vw/core/learner.h (hunk 2 above)
Comment (bassmang) [reply vowpalwabbit/core/include/vw/core/learner.h]: updated to lowercase

File: vowpalwabbit/core/src/reductions/search/search.cc [hunk 6]
```diff
@@ -3366,6 +3366,9 @@ base_learner* VW::reductions::search_setup(VW::setup_base_i& stack_builder)
 
   cdbg << "num_learners = " << priv.num_learners << endl;
 
+  size_t ws = 1;
+  if (task_string == "sequencespan" || task_string == "multiclasstask") { ws = 2; }
```
Comment (lalo) [inline vowpalwabbit/core/src/reductions/search/search.cc:3370]: can you add the assert to search.cc just like the other PR

Comment (lalo) [reply vowpalwabbit/core/src/reductions/search/search.cc:3370]: also add a comment here referencing what is going on and mention the assert that fails

Comment (lalo) [reply vowpalwabbit/core/src/reductions/search/search.cc:3370]: https://github.com/VowpalWabbit/vowpal_wabbit/pull/4031

Comment (bassmang) [reply vowpalwabbit/core/src/reductions/search/search.cc]: updated this fix so removed this

File: cs/unittest/TestSearch.cs [hunk 7]
```diff
@@ -145,10 +145,10 @@ void RunSearchPredict(Multiline multiline)
                     driver.ForEachMultiline(rawvw.Learn);
 
                     rawvw.EndOfPass();
-                    driver.Reset();
+                    //driver.Reset(); -- Throws if search is called on out of bound learner
```
Comment (lalo) [inline cs/unittest/TestSearch.cs:148]: we dont need this anymore with your patch right?

File: test/train-sets/ref/automl_readable_qc.txt [hunk 8]
```diff
@@ -1801,7 +1791,7 @@ event_sum 0
 action_sum 0
 :1
 initial_t 0
-norm normalizer 0
+norm normalizer 98.9861
```
Comment (lalo) [inline test/train-sets/ref/automl_readable_qc.txt:1794]: since we are already breaking the model, we could also remove this norm and total_weight since they are duplicate. Thoughts?

File: test/train-sets/ref/houses.stdout [hunk 9]
```diff
@@ -1,30 +1,30 @@
 0
-HiddenBias:29015:1:0@0
+	HiddenBias:29015:1:0@0
```
Comment (lalo) [inline test/train-sets/ref/houses.stdout:2]: @jackgerrits what triggered this change? I feel this was from an old change of formatting you did a while back

File: test/train-sets/same_model_test.1.dat [hunk 10]
```diff
@@ -1,2 +1,2 @@
-1.0 |a short |b good
--1.0 |a median
+1.0 |a short small |b good
```
Comment (lalo) [inline test/train-sets/same_model_test.1.dat:1]: why did you change this file?

File: cs/unittest/TestSearch.cs (hunk 7 above)
Comment (bassmang) [reply cs/unittest/TestSearch.cs:148]: I think this fails, I can check again

File: vowpalwabbit/core/src/reductions/gd.cc [hunk 11]
```diff
@@ -1504,6 +1581,7 @@ base_learner* VW::reductions::gd_setup(VW::setup_base_i& stack_builder)
           .set_merge_with_all(::merge)
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/gd.cc:1581]: Please ensure all merge, add and subtract logic is updated

File: cs/unittest/TestSearch.cs (hunk 7 above)
Comment (lalo) [reply cs/unittest/TestSearch.cs:148]: @lokitoth do you have any theory on why this change would break this test?

File: test/pred-sets/ref/ccb_quad.inv [hunk 12]
```diff
@@ -30,6 +30,9 @@ sd::oec.weighted_labeled_examples 8
 current_pass 1
 l1_state 0
 l2_state 1
+gd_ppw_state.size() = 1
```
Comment (lalo) [inline test/pred-sets/ref/ccb_quad.inv:33]: we might be able to avoid breaking compat, if we put this extra state after the weights and not before

File: vowpalwabbit/core/src/reductions/gd.cc [hunk 13]
```diff
@@ -825,7 +852,12 @@ void NO_SANITIZE_UNDEFINED learn(VW::reductions::gd& g, base_learner& base, VW::
   assert(ec.l.simple.label != FLT_MAX);
   assert(ec.weight > 0.);
   g.predict(g, base, ec);
+  // this state should only matter on learn and not predict
+  // assert(g.current_model_state == nullptr);
+  g.current_model_state = &(g.per_model_states.at(ec.ft_offset / g.all->weights.stride()));
```
Comment (lalo) [inline vowpalwabbit/core/src/reductions/gd.cc:857]: I think we need to use the VW::scope_exit(

File: vowpalwabbit/core/src/reductions/gd.cc [hunk 14]
```diff
@@ -312,6 +314,21 @@ void subtract(const VW::workspace& ws1, const VW::reductions::gd& data1, const V
   }
 }
 
+void resize_ppw_state(VW::reductions::gd& g, size_t factor, size_t /* max_ft_offset */)
```
Comment (lalo) [inline vowpalwabbit/core/src/reductions/gd.cc:317]: @jackgerrits

File: test/train-sets/ref/automl_readable_qc.txt (hunk 8 above)
Comment (bassmang) [reply test/train-sets/ref/automl_readable_qc.txt:1794]: we should move this to next PR which can break model

File: test/train-sets/same_model_test.1.dat (hunk 10 above)
Comment (bassmang) [reply test/train-sets/same_model_test.1.dat:1]: this was need to get the action/event sums the same now that they are stored per model

File: vowpalwabbit/core/src/reductions/gd.cc (hunk 13 above)
Comment (bassmang) [reply vowpalwabbit/core/src/reductions/gd.cc:867]: done

**Comment by lalo** [comment]
@bassmang shouldn't loss calculation also be per ft_offset? we might be reporting wrong loss in eps-decay a loss that is being affected by all live models? thoughts?

File: vowpalwabbit/core/include/vw/core/global_data.h [hunk 15]
```diff
@@ -188,6 +188,7 @@ class workspace
   void* /*Search::search*/ searchstr;
 
   uint32_t wpp;
+  std::vector<std::pair<uint32_t, size_t>> wpp_innermost_removals;
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/global_data.h:191]: Can this be removed?

File: vowpalwabbit/core/include/vw/core/learner.h [hunk 16]
```diff
@@ -98,6 +98,7 @@ void debug_increment_depth(polymorphic_ex ex);
 void debug_decrement_depth(polymorphic_ex ex);
 void increment_offset(polymorphic_ex ex, const size_t increment, const size_t i);
 void decrement_offset(polymorphic_ex ex, const size_t increment, const size_t i);
+size_t get_offset(polymorphic_ex ex);
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/learner.h:101]: Can this be removed?

File: vowpalwabbit/core/include/vw/core/reductions/epsilon_decay.h [hunk 17]
```diff
@@ -57,11 +57,8 @@ class epsilon_decay_data
   bool _predict_only_model;
 
   // TODO: delete all this, gd and cb_adf must respect ft_offset, see header import of automl.cc
-  std::vector<double> per_live_model_state_double;
+  LEARNER::learner* adf_learner = nullptr;  //  re-use print from cb_explore_adf
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/reductions/epsilon_decay.h:60]: Is this change related to the ppw change?

File: vowpalwabbit/core/include/vw/core/vw_versions.h [hunk 18]
```diff
@@ -63,6 +63,9 @@ constexpr VW::version_struct VERSION_FILE_WITH_ACTIVE_SEEN_LABELS{9, 0, 0};
 /// Moved option values from command line to model data
 constexpr VW::version_struct VERSION_FILE_WITH_L1_AND_L2_STATE_IN_MODEL_DATA{9, 0, 0};
 
+/// GD, freegrad, ftrl are able to persist the correct state if ppw is higher than 1
+constexpr VW::version_struct VERSION_FILE_WITH_GD_PPW_STATE{9, 7, 0};
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/vw_versions.h:67]: Can this be removed?

File: vowpalwabbit/core/src/learner.cc [hunk 19]
```diff
@@ -756,3 +756,9 @@ void VW::LEARNER::details::decrement_offset(polymorphic_ex ex, const size_t incr
   }
   debug_decrement_depth(ex);
 }
+
+size_t VW::LEARNER::details::get_offset(polymorphic_ex ex)
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/learner.cc:760]: Can this be removed?

File: vowpalwabbit/core/src/reductions/automl.cc [hunk 20]
```diff
@@ -115,6 +111,7 @@ void pre_save_load_automl(VW::workspace& all, automl<CMType>& data)
   }
   options.insert("interactions", "");
   options.get_typed_option<std::vector<std::string>>("interactions").value(interactions_opt);
+  all.wpp_innermost_removals.push_back(std::make_pair(data.cm->max_live_configs, 0));
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/automl.cc:114]: Can this be removed?

File: vowpalwabbit/core/src/reductions/cb/cb_algs.cc [hunk 21]
```diff
@@ -166,6 +166,7 @@ std::shared_ptr<VW::LEARNER::learner> VW::reductions::cb_algs_setup(VW::setup_ba
 
   VW::details::cb_to_cs& c = data->cbcs;
 
+  // todo: is this being used?
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/cb/cb_algs.cc:169]: Can this be removed?

File: vowpalwabbit/core/src/reductions/gd.cc [hunk 22]
```diff
@@ -838,7 +865,18 @@ void learn(VW::reductions::gd& g, VW::example& ec)
...
-  update<sparse_l2, invariant, sqrt_rate, feature_mask_off, adax, adaptive, normalized, spare>(g, ec);
+  // this state should only matter on learn and not predict
+  // assert(g.current_model_state == nullptr);
+  // Guard example state restore against throws
+  auto restore_guard = VW::scope_exit(
+      [&g, &ec]
+      {
+        g.current_model_state = &(g.per_model_states.at(ec.ft_offset / g.all->weights.stride()));
+
+        update<sparse_l2, invariant, sqrt_rate, feature_mask_off, adax, adaptive, normalized, spare>(g, ec);
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/gd.cc:876]: Can update throw? What happens then?

File: vowpalwabbit/core/src/reductions/gd.cc [hunk 23]
```diff
@@ -1200,6 +1240,16 @@ void VW::details::save_load_online_state_gd(VW::workspace& all, VW::io_buf& mode
...
+  /*if (!read || all.model_file_ver >= VW::version_definitions::VERSION_FILE_WITH_GD_PPW_STATE)
+  {
+    if (read)
+    {
+      //pms.clear();
+      VW::model_utils::read_model_field(model_file, pms[0]);
+    }
+    else { VW::model_utils::write_model_field(model_file, pms[0], "gd_ppw_state", text); }
+  }*/
+
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/gd.cc:1252]: remove commented out code

File: vowpalwabbit/core/src/reductions/gd.cc [hunk 24]
```diff
@@ -1267,11 +1317,32 @@ void save_load(VW::reductions::gd& g, VW::io_buf& model_file, bool read, bool te
...
+          std::vector<VW::reductions::details::per_model_state> new_pms;
+          for (auto idx = 0; idx < all.wpp; idx += removals.first)
+          {
+            new_pms.push_back(g.per_model_states[idx + removals.second]);
+          }
+          g.per_model_states = std::move(new_pms);
+        }
+        VW::model_utils::write_model_field(model_file, g.per_model_states[0], "_gd_ppw_state", text);
+      }*/
+
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/gd.cc:1345]: remove commented out code

File: vowpalwabbit/core/src/reductions/search/search_multiclasstask.cc [hunk 25]
```diff
@@ -24,7 +24,10 @@ void initialize(Search::search& sch, size_t& num_actions, VW::config::options_i&
 {
   task_data* my_task_data = new task_data();
   sch.set_options(0);
-  sch.set_num_learners(num_actions);
+  size_t num_learner_ceil = 1;
+  while (num_learner_ceil < num_actions) { num_learner_ceil *= 2; }
+  // TODO: temporary fix to ensure enough learners, look into case where learner is created outside bounds
+  sch.set_num_learners(num_learner_ceil);
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/search/search_multiclasstask.cc:30]: This makes me nervous...

File: vowpalwabbit/core/include/vw/core/global_data.h (hunk 15 above)
Comment (bassmang) [reply vowpalwabbit/core/include/vw/core/global_data.h:191]: removed

File: vowpalwabbit/core/include/vw/core/learner.h (hunk 16 above)
Comment (bassmang) [reply vowpalwabbit/core/include/vw/core/learner.h:101]: removed

File: vowpalwabbit/core/include/vw/core/reductions/epsilon_decay.h (hunk 17 above)
Comment (bassmang) [reply vowpalwabbit/core/include/vw/core/reductions/epsilon_decay.h:60]: removed

File: vowpalwabbit/core/include/vw/core/vw_versions.h (hunk 18 above)
Comment (bassmang) [reply vowpalwabbit/core/include/vw/core/vw_versions.h:67]: removed

File: vowpalwabbit/core/src/learner.cc (hunk 19 above)
Comment (bassmang) [reply vowpalwabbit/core/src/learner.cc:760]: removed

File: vowpalwabbit/core/src/reductions/automl.cc (hunk 20 above)
Comment (bassmang) [reply vowpalwabbit/core/src/reductions/automl.cc:114]: removed

File: vowpalwabbit/core/src/reductions/cb/cb_algs.cc (hunk 21 above)
Comment (bassmang) [reply vowpalwabbit/core/src/reductions/cb/cb_algs.cc:169]: removed

File: vowpalwabbit/core/src/reductions/search/search_multiclasstask.cc (hunk 25 above)
Comment (bassmang) [reply vowpalwabbit/core/src/reductions/search/search_multiclasstask.cc:30]: emailed john about this with 5 or so options, agreed this is the most suitable solution

File: vowpalwabbit/core/src/reductions/gd.cc (hunk 22 above)
Comment (bassmang) [reply vowpalwabbit/core/src/reductions/gd.cc:876]: why would update throw?

File: vowpalwabbit/core/src/reductions/gd.cc [hunk 26]
```diff
@@ -838,7 +865,18 @@ void learn(VW::reductions::gd& g, VW::example& ec)
...
+  // Guard example state restore against throws
+  auto restore_guard = VW::scope_exit(
+      [&g, &ec]
+      {
+        g.current_model_state = &(g.per_model_states.at(ec.ft_offset / g.all->weights.stride()));
+
+        update<sparse_l2, invariant, sqrt_rate, feature_mask_off, adax, adaptive, normalized, spare>(g, ec);
+        assert(g.current_model_state == nullptr);  // update clears this pointer
+        g.current_model_state = nullptr;
+      });
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/gd.cc:879]:
```suggestion
  g.current_model_state = &(g.per_model_states.at(ec.ft_offset / g.all->weights.stride()));
  update<sparse_l2, invariant, sqrt_rate, feature_mask_off, adax, adaptive, normalized, spare>(g, ec);
  assert(g.current_model_state == nullptr);  // update clears this pointer
  auto restore_guard = VW::scope_exit(
      [&g]
      {
        g.current_model_state = nullptr;
      });
```

File: test/test-sets/ref/active_cover.stderr [hunk 27]
```diff
@@ -5,7 +5,7 @@ Reading datafile = test-sets/rcv1_small_test.data
 num sources = 1
 Num weight bits = 18
 learning rate = 0.5
-initial_t = 217.709
+initial_t = 222.626
```
Comment (jackgerrits) [inline test/test-sets/ref/active_cover.stderr:8]: Why did this change?

File: test/train-sets/ref/lrqfa_nn_pred_rev.stderr [hunk 28]
```diff
@@ -12,15 +12,15 @@ Input label = SIMPLE
...
+0.110677 0.110677            1            1.0        -1.0000         0.3391        6
+0.300991 0.491306            2            2.0         1.0000         0.5742        6
+0.273194 0.245398            4            4.0         1.0000         0.5799        6
 
 finished run
 number of examples = 4
 weighted example sum = 4.000000
 weighted label sum = 0.000000
-average loss = 0.412701
+average loss = 0.273194
```
Comment (jackgerrits) [inline test/train-sets/ref/lrqfa_nn_pred_rev.stderr:23]: This change is quite large

File: test/train-sets/ref/spin_off_aml_save.stderr [hunk 29]
```diff
@@ -2,7 +2,7 @@ final_regressor = models/aml_spin_off.m
 using no cache
 Reading datafile = train-sets/automl_spin_off.txt
 num sources = 1
-Num weight bits = 18
+Num weight bits = 19
```
Comment (jackgerrits) [inline test/train-sets/ref/spin_off_aml_save.stderr:5]: Why did the num bits change?

File: test/core.vwtest.json [hunk 30]
```diff
@@ -5476,7 +5476,7 @@
   {
     "id": 425,
     "desc": "Spin off automl model to lower bitsize to cb_explore",
-    "vw_command": "--cb_explore_adf -d train-sets/automl_spin_off.txt --noconstant -f models/aml_spin_off.m -b 18 --predict_only_model --automl 2",
+    "vw_command": "--cb_explore_adf -d train-sets/automl_spin_off.txt --noconstant -f models/aml_spin_off.m -b 19 --predict_only_model --automl 4",
```
Comment (jackgerrits) [inline test/core.vwtest.json:5479]: Why change the test in this PR?

File: test/train-sets/ref/active_cover.stderr [hunk 31]
```diff
@@ -17,15 +17,15 @@ loss     last          counter         weight          label        predict feat
...
+0.281479 0.203259          211          174.7        -1.0000        -1.0000      119
 
 finished run
 number of examples = 250
-weighted example sum = 217.709346
-weighted label sum = 15.483050
-average loss = 0.264386
-best constant = 0.142477
-best constant's loss = 0.690616
+weighted example sum = 222.625671
```
Comment (jackgerrits) [inline test/train-sets/ref/active_cover.stderr:25]: What are the reasons for these changes?

File: test/train-sets/ref/spin_off_aml_save.stderr (hunk 29 above)
Comment (bassmang) [reply test/train-sets/ref/spin_off_aml_save.stderr:5]: I updated this test to ensure it works for more models/higher bitsizes. Pretty unrelated I can revert

File: test/core.vwtest.json (hunk 30 above)
Comment (bassmang) [reply test/core.vwtest.json:5479]: same as above

File: test/train-sets/ref/lrqfa_nn_pred_rev.stderr (hunk 28 above)
Comment (lalo) [reply test/train-sets/ref/lrqfa_nn_pred_rev.stderr:23]: but its moving to the right direction

File: vowpalwabbit/core/src/reductions/gd.cc [hunk 32]
```diff
@@ -1503,6 +1561,8 @@ std::shared_ptr<VW::LEARNER::learner> VW::reductions::gd_setup(VW::setup_base_i&
 
   all.weights.stride_shift(static_cast<uint32_t>(::ceil_log_2(stride - 1)));
 
+  resize_ppw_state(*g, ppw);
```
Comment (lalo) [inline vowpalwabbit/core/src/reductions/gd.cc:1564]: make this logic part of the constructor in line 1428 - now that we have that info so there's no need to resize

File: vowpalwabbit/core/include/vw/core/cached_learner.h [hunk 33]
```diff
@@ -11,7 +11,7 @@ namespace VW
 class cached_learner : public setup_base_i
 {
 public:
-  std::shared_ptr<VW::LEARNER::learner> setup_base_learner() override { return _cached; }
+  std::shared_ptr<VW::LEARNER::learner> setup_base_learner(size_t) override { return _cached; }
```
[the remark this answers is not in the document]
Comment (lalo) [reply vowpalwabbit/core/include/vw/core/cached_learner.h:14]: cached_learner currently doesnt exercise this api so its a moot point

File: test/train-sets/ref/active_cover.stderr (hunk 31 above)
Comment (lalo) [reply test/train-sets/ref/active_cover.stderr:25]: loss went down

**Comment by lalo** [comment]
LGTM

**Review by bassmang — APPROVED** [review]
